docs/architecture/adr-d001-inter-validator-set-messaging.md

```diff
@@ -0,0 +1,222 @@
+Validators) connected to all other Validators. In an extreme case, a Validator node can be connected only to Full
+nodes, effectively blocking its communication and excluding it from participation in the consensus.
+
+### Solution
+
+Each Validator node shall allocate a preconfigured number of connectivity slots for communication with other Validators
+that belong to the same Validator Set.
+
+In the current implementation, the ABCI application manages the rotation of Validator Sets. When rotation is needed, 
+the application sends information about the new Validator Set in response to the `EndBlock` message. 
+
+`EndBlock` message shall be extended by adding a network address of each Validator node. Based on that, each
+member of the active Validator Set will connect with a predefined number of Peer Validators belonging to the same 
+Validator Set. A list of Peer Validators will be determined using the algorithm described in 
+[DIP 0006](https://github.com/dashpay/dips/blob/master/dip-0006.md#building-the-set-of-deterministic-connections).
+
+This solution introduces a risk of delayed propagation of new block due to limited connection slots between Full and
+Validator nodes. To mitigate this risk, we can tune the number of Full node connection slots independently from
+Validator node connection slots.
+
+## Alternative Approaches
+
+### Rotate on Tenderdash
+
+ABCI application can send the whole master node list and quorum list to Tenderdash and
+potentially rotate on its side. However, it would split the responsibility for the
+management of Validator Set rotation between multiple components. It would also require
+much more additional implementation. It would be better to keep the current Tendermint
+design and concept being a logic-less consensus library and keep all business logic in
+the ABCI application.
+
+### TODO
+
+## Decision
+
+> TODO: This section records the decision.
+> It is best to record as much info as possible from the discussion that happened. This aids in not having to go back
+> to the Pull Request to get the needed information.
+
+## Detailed Design
+
+### What are the user requirements?
+
+Having a network with much more Full nodes than Validators ensure each Validator Node can connect (directly or
+through another Validator) to any other Validator which is a member of the same active Validator Set.
+
+### What systems will be affected?
+
+* Tenderdash
+* ABCI App
+
+### What new data structures are needed, what data structures need changes?
+
+1. Separate Validator Set member connection pool from Full node connection pool
+2. Additional tuning parameters:
+    1. Full node connection slots
+    2. Validator node connection slots
+
+### What new APIs will be needed, what APIs will change?
+
+In the ABCI protocol, we add a network address ( `tendermint.p2p. NetAddress` ) of each member of the active Validator
```

Comment:
Do we need a validator node id for the p2p layer or IP will be enough?

Reply:
We do need the node ID, as it's in use internally by Tenderdash when establishing connections. To be discussed if it should be sent by the ABCI App, or generated on Tenderdash.

It's [easy to calculate](https://github.com/dashevo/tenderdash/blob/7bb6746cee3ed6cb302855656cd2634a100d6ed0/p2p/key.go#L46), assuming we have the public key (and we have it, as it's part of ValidatorUpdate).

Reply:
Actually, after some thinking, I believe NetAddress struct should NOT contain ID. It should be just a generic NetAddress type. I will create a separate type that will only contain IP address and port (+ support for ipv6).

Reply:
What about this?

```proto3
enum NetAddressType {
  IPv4 = 0;
};

message NetAddress {
  reserved 1; // For compatibility with tendermint.p2p.NetAddress, we reserve ID (1) field
  reserved "id";
  string address = 2;
  uint32 port = 3;
  NetAddressType address_type = 4;
 }
 ```

Reply:
the address should not be a string imo, since it's really a uint128 number.